From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Daniel Micay <daniel.micay@grapheneos.org>
Date: Tue, 8 Apr 2025 22:04:15 -0400
Subject: [PATCH 07/31] fix upstream RecoverySystem.verifyPackage vulnerability

This was not directly exploitable due to there being 2 layers of update
package signature verification and downgrade protection, but the first
layer of protection should work properly to avoid a vulnerability in the
2nd layer being exploited.
---
 core/java/android/os/RecoverySystem.java | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/core/java/android/os/RecoverySystem.java b/core/java/android/os/RecoverySystem.java
index de995527ee..7ab53d3422 100644
--- a/core/java/android/os/RecoverySystem.java
+++ b/core/java/android/os/RecoverySystem.java
@@ -334,6 +334,9 @@ public class RecoverySystem {
             if (certificates == null || certificates.length == 0) {
                 throw new SignatureException("signature contains no certificates");
             }
+            if (certificates.length > 1) {
+                throw new SignatureException("signature contains multiple certificates");
+            }
             X509Certificate cert = certificates[0];
             PublicKey signatureKey = cert.getPublicKey();
 
@@ -341,6 +344,9 @@ public class RecoverySystem {
             if (signerInfos == null || signerInfos.length == 0) {
                 throw new SignatureException("signature contains no signedData");
             }
+            if (signerInfos.length > 1) {
+                throw new SignatureException("signature contains multiple signedData");
+            }
             SignerInfo signerInfo = signerInfos[0];
 
             // Check that the public key of the certificate contained
